app/Jobs/GenerateSimilarityIndex.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace Biigle\Jobs;
+
+
+use Biigle\Image;
+use Biigle\Jobs\Job;
+use Biigle\Volume;
+use Storage;
+use File;
+use Exception;
+use Log;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Queue\SerializesModels;
+
+
+class GenerateSimilarityIndex extends Job implements ShouldQueue
+{
+    use SerializesModels, InteractsWithQueue;
+
+    /**
+     * The number of times the job may be attempted.
+     *
+     * @var int
+     */
+    public $tries = 2;
+
+    /**
+     * Ignore this job if the image does not exist any more.
+     *
+     * @var bool
+     */
+    protected $deleteWhenMissingModels = true;
+
+
+    /**
+     * The volume to process
+     *
+     * @var Volume
+     */
+    public $volume;
+
+    /**
+     * The hash values according to the images in volume
+     *
+     * @var array
+     */
+    protected $hashValues;
+
+    /**
+     * The similarity values according to the images in volume
+     *
+     * @var array
+     */
+    protected $similarityIndices;
+
+    /**
+     * Create a new job instance.
+     *
+     * @param Image $image The image to generate process.
+     *
+     * @return void
+     */
+    public function __construct(Volume $volume)
+    {
+        $this->volume = $volume;
+        $this->hashValues = [];
+        $this->similarityIndices = [];
+
+    }
+
+    /**
+     * Execute the job.
+     *
+     * @return void
+     * @throws Exception
+     */
+    public function handle()
+    {
+       $imageHashArray = $this->createImageHashArray();
+        if ($imageHashArray !== false) {
+            $similarityIndexArray = $this->createSimilarityIndexArray($imageHashArray);
+            if (!is_null($similarityIndexArray)) {
+                foreach ($this->volume->images as $image) {
+                    $id = $image->id;
+                    $similarityIndex = $similarityIndexArray[$id];
+                    $image->similarityIndex = $similarityIndex;
+                    $image->save();
+                }
+            }
+        }
+
+
+
+    }
+
+    /**
+     * Creates an array with all the ids and hash values for each image
+     * If one image has no hash value the function will return false for no more doing this job.
+     *
+     * @return array|false
+     */
+    public function createImageHashArray()
+    {
+        $imageHashArray = array();
+
+        foreach ($this->volume->images as $image) {
+            $id = $image->id;
+            $hash = $image->hash;
+            if (is_null($hash))
+            {
+               return false;
+            }
+            $imageHashArray[$id] = $hash;
+        }
+        return $imageHashArray;
+
```

Comment:
Maybe something like:

```php
if ($this->volume->images()->whereNull('hash')->exists()) {
  return false;
}

return $this->volume->images()->pluck('hash', 'id');
```
**But:** This is probably no huge performance gain, as you still fetch all images for the loop in line 85. Maybe it's even more efficient in the current implementation. Instead of a database query (`->images()`) you could also work on the whole [collection](https://laravel.com/docs/6.x/collections) of images (`->images`):


```php
if ($this->volume->images->whereNull('hash')->isNotEmpty()) {
  return false;
}

return $this->volume->images->pluck('hash', 'id');
```
The collection is fetched only once from the database (Laravel takes care of this) and cached for subsequent use (e.g. in line 85). This way you have only one database query instead of three and still fewer lines of code.  :wink:

Reply:
I implement the version with the collection, good advise thanks